retrievalmarket/network/libp2p_impl.go

```diff
@@ -107,8 +107,11 @@ func (impl *libp2pRetrievalMarketNetwork) openStream(ctx context.Context, id pee
 		if nAttempts == impl.maxStreamOpenAttempts {
 			return nil, xerrors.Errorf("exhausted %d attempts but failed to open stream, err: %w", int(impl.maxStreamOpenAttempts), err)
 		}
-		d := b.Duration()
-		time.Sleep(d)
+		select {
+		case <-ctx.Done():
+			return nil, xerrors.Errorf("backoff canceled by context")
+		case <-time.After(b.Duration()):
```

Comment:
Ideally use `time.NewTimer` to avoid (temporarily) [leaking](https://golang.org/pkg/time/#After) a timer on cancellation. But this isn't a huge deal.

Reply:
Yeah, considering that but wouldn't imagine would be a big issue having the timers hanging around a bit. 
I'll change to cancel them.